### PR TITLE
uio: fix readv & writev return value, set EINVAL errno

### PR DESCRIPTION
**Pull request by ziemleszcz**
According to POSIX:

1. If a read() is interrupted by a signal before it reads any
data, it shall return -1 with errno set to [EINTR].
2. If a read() is interrupted by a signal after it has
successfully read some data, it shall return the number of
bytes read.
3. If write() is interrupted by a signal before it writes any
data, it shall return -1 with errno set to [EINTR].
4. If write() is interrupted by a signal after it successfully
writes some data, it shall return the number of bytes written.

JIRA: RTOS-46

## Motivation and Context

Current implementation of readv/writev breaks dropbear (channel write).

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: sys/uio.c [hunk 1]
```diff
@@ -42,16 +42,21 @@ ssize_t readv(int fd, const struct iovec *iov, int iovcnt)
...
+	}
 
 	/* TODO: file locking */
 
 	for (i = 0; i < iovcnt; ++i) {
 		if (iov[i].iov_len > 0) {
 			n = read(fd, iov[i].iov_base, iov[i].iov_len);
-			if (n == -1)
+			if (n < 0) {
+				if ((tot_len > 0 && errno == EINTR) || errno == EAGAIN)
```
Comment (agkaminski) [inline sys/uio.c:56]: Shouldn't we return `-1` in case of `EAGAIN`?

File: sys/uio.c [hunk 2]
```diff
@@ -69,16 +74,21 @@ ssize_t writev(int fd, const struct iovec *iov, int iovcnt)
...
+	}
 
 	/* TODO: file locking */
 
 	for (i = 0; i < iovcnt; ++i) {
 		if (iov[i].iov_len > 0) {
 			n = write(fd, iov[i].iov_base, iov[i].iov_len);
-			if (n == -1)
+			if (n < 0) {
+				if ((tot_len > 0 && errno == EINTR) || errno == EAGAIN)
```
Comment (agkaminski) [inline sys/uio.c:88]: As above, shouldn't we return `-1` in case of `EAGAIN`?

**Comment by agkaminski** [comment]
Add reviewers as reviewers, not assignees.

File: sys/uio.c (hunk 2 above)
Comment (ziemleszcz) [reply sys/uio.c:88]: Yes if tot_len is zero. If tot_len is non-zero we should return number of bytes written and return EAGAIN on next call. For context check out this code from dropbear:

	written = writev(fd, iov, io_count);

	if (written < 0) {
		if (errno != EINTR && errno != EAGAIN) {
			TRACE(("channel IO write error fd %d %s", fd, strerror(errno)))
			close_chan_fd(channel, fd, SHUT_WR);
			return DROPBEAR_FAILURE;
		}
	} else {
		int cbuf_written = MIN(circ_len1+circ_len2, (unsigned int)written);
		cbuf_incrread(cbuf, cbuf_written);
		if (morelen) {
			*morelen = written - cbuf_written;
		}
		channel->recvdonelen += written;
	}

Comment (agkaminski) [reply sys/uio.c:88]: Ok, so your condition should look like that:
`if (tot_len > 0 && (errno == EINTR || errno == EAGAIN))`
i.e. you should return `tot_len` only, if it's not zero.

**Review by agkaminski — APPROVED** [review]

**Pull request merged**